### DISCUSSION TITLE
RAMP as a monorepo

### DISCUSSION THREAD

**Discussion by AleksueiR** [fgpv-vpgf/fgpv-vpgf#4118] [Team Posts]
We discussed bringing the plugins repo inside the main one as a Git submodule, but it never amounted to anything since the build toolchains in these repos are different which is a problem. The main reason for wanting to keep all plugin code in with the base is the painful experience of developing plugins using the existing setup (a need for two PRs, keeping a built version of RAMP inside the plugins repo, and other contortions).

Last week I stumbled on an article on how to set up a monorepo (https://areknawo.com/full-blown-monorepo-setup-walkthrough/) using `Lerna` package manager (https://lerna.js.org/).

> **Monorepo** is a term referring to the way you organize your code-base within a single repository.

> **Lerna** is a tool that optimizes the workflow around managing multi-package repositories with git and npm.

This seems like a potential solution where we can bring the plugins repo (and potentially geoApi and geoSearch if we want to) inside the main RAMP repo keeping their respective toolchains separate with `Lerna`. Apparently, there are many libraries that follow this route and keep separate (but related - ones that makes sense to group) packages under the same repo (like `Babel`, `Vue-cli`, and `create-react-app`).

Thoughts?

**Comment by james-rae** [comment]
![monorepo](https://user-images.githubusercontent.com/7585245/63354892-3ca97e00-c333-11e9-9fe3-2a2659f808c8.png)

**Comment by milespetrov** [comment]
Lerna sounds interesting! I think it will help us better manage the various code bases we have.  I just wonder how painful this will be to implement... some questions that come to mind:

- Will the various build processes be merged into one? i.e integrate the plugin build and the geosearch builds into the ramp build pipeline? Or will we continue to have separate build processes within the various "subfolders"?
- Will the conflicting build environments (plugins target `es5` and compile to `commonJS` via TypeScript, ramp is mostly Babel `'env', 'stage-2'` with some TypeScript mixed in - targeting ES6) be united?
- Will all pushes to the ramp repo trigger npm publishes for updated subrepos? Will publishing these subrepos trigger a new version/publish of RAMP? See https://github.com/lerna/lerna#fixedlocked-mode-default

These questions don't need to be answered here, just thought food. There may be more gotchas, if I think of any I'll update here.

**Comment by james-rae** [comment]
The tech is beyond me but if we can crush all the grousy issues surrounding the separate plugin repo, I'd say this is worth some time to investigate.

geoApi is peachy at the moment. The only chore is pushing the odd tag, and tags are fun.  If moving the repo into a subrepo of sorts makes things easier and doesn't require any significant changes to geoApi, I have no issues.

**Comment by AleksueiR** [comment]
Here's a POC monorepo (without any RAMP code for the moment) with two packages using different build processes and hot-reload on both packages: https://github.com/muffin-crusaders/monorepo-ramp-test

@milesap 
> Will the various build processes be merged into one?

They could be, but this is not required as packages' respective build processes will still work.  During future refactors, we can keep an eye on trying to consolidate the build processes though.

> Will the conflicting build environments (plugins target es5 and compile to commonJS via TypeScript, ramp is mostly Babel 'env', 'stage-2' with some TypeScript mixed in - targeting ES6) be united?

Same as above. We don't have to unite anything right now (except for moving all the code into the core repo) and different builds should work as before using whatever env vars we use right now.

> Will all pushes to the ramp repo trigger npm publishes for updated subrepos?

I no longer think Lerna is the best option ([see here why](https://github.com/muffin-crusaders/monorepo-ramp-test#notes)) and was experimenting with [Rush](https://rushjs.io). With Rush, we don't have to publish anything to NPM at all as the packages are handles locally. We [can publish](https://rushjs.io/pages/commands/rush_publish/) if we want to, but it's not required for the monorepo to work.

@james-rae 
> if we can crush all the grousy issues surrounding the separate plugin repo

Yes, this will solve multiple-PR issues.

> If moving the repo into a subrepo of sorts makes things easier and doesn't require any significant changes to geoApi,

It shouldn't cause any problems as the build processes stay separate and RAMP core would be just importing the geoApi build as usual.

**Comment by AleksueiR** [comment]
A tool for the multi-to-mono transition: https://github.com/unravelin/tomono

**Comment by AleksueiR** [comment]
More tools: https://github.com/shopsys/shopsys/tree/master/packages/monorepo-tools

`tomono` seems to have a limitation since all packages need to be merged at the same time; you can't merge additional packages later.
`monorepo-tools` seems to be an order of magnitude slower than `tomono` though.

**Comment by AleksueiR** [comment]
## Update: It's alive 🧟‍♀️

I have another POC repo to try out: https://github.com/muffin-crusaders/monorepo-ramp-test-2

It includes all the RAMP packages (`geoapi`, `core`, `plugins`, and `geosearch`), but only `geoapi` is cross-linked to `ramp` right now. Read the readme for details.

**Comment by AleksueiR** [comment]
Belated update:
- wired table and custom export plugins into the monorepo setup
  - hot-reload works for both plugins and for RAMP Core
  - plugins are split into their own separate packages, so each one can haz their own dependencies and build steps
- gave plugins new (funny) names

Hopefully, will finish wiring the rest of the plugins and geosearch package into the setup and then everyone can try it out.

**Comment by AleksueiR** [comment]
Milestone achieved: stuff works, no tests. 🥳🤪🎂
Please checkout the [monoRAMP](https://github.com/muffin-crusaders/monorepo-ramp-test-2), read the `Getting Started` section, and, please, try it on your machine to see if it works.

**Comment by RyanCoulsonCA** [comment]
🏆 WOMM, but I had to lower the `nodeSupportedVersionRange` to `10.10.0` because that's what version of node my laptop is running. Looks like I'll need admin permissions to update.

**Comment by AleksueiR** [comment]
@RyanCoulsonCA that's awesome! Node version used by Rush doesn't seem like a hard requirement, so it should be fine even if you won't update.
